src/Connector/CachingConnector.php

```diff
@@ -74,8 +87,32 @@ public function isCacheEnabled()
         return $this->cacheEnabled;
     }
 
-    private function hash(array $structure)
+    public function getCacheKeyGenerator()
+    {
+        return $this->cacheKeyGenerator;
+    }
+
+    public function setCacheKeyGenerator(CacheKeyGenerator $cacheKeyGenerator)
     {
-        return str_replace(str_split('{}()/\@:'), '.', json_encode($structure, JSON_UNESCAPED_SLASHES));
+        $this->cacheKeyGenerator = $cacheKeyGenerator;
+    }
+
+    /**
+     * @param mixed $key
+     *
+     * @return string
+     *
+     * @throws InvalidCacheKeyException
+     */
+    private function validateCacheKey($key)
+    {
+        if (!is_string($key)) {
+            throw new InvalidCacheKeyException('Cache key must be of type string.');
+        }
+        if (strpbrk($key, '{}()/\@:') !== false) {
```

Comment:
Consider defining this as a constant, e.g. `const RESERVED_CHARACTERS`. It could be reused in tests and the following exception.